### PR TITLE
ARROW-14168: [R] Warn only once about arrow function differences

### PR DESCRIPTION
**Pull request by alistaire47**
Addresses [ARROW-14168](https://issues.apache.org/jira/browse/ARROW-14168) by changing `median()` and `quantile()` to warn once, and adjusts the tests accordingly.

### CODE REVIEW

File: r/tests/testthat/test-dplyr-summarize.R
Comment (alistaire47) [inline r/tests/testthat/test-dplyr-summarize.R:397]: These big diffs are just removing `expect_warning()` and dedenting everything accordingly

File: r/tests/testthat/test-dplyr-summarize.R
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-summarize.R:324]: Hmm, this will quiet all rlang-type warnings, which is probably not quite what we want. Maybe this should be `"default"` instead of `"quiet"`? Though, this is moot if we go in the other direction up above.

**Review by jonkeane — CHANGES_REQUESTED** [review]
Thanks for pushing on this! Turns out this ends up pulling on some threads in our testing setup.

I wonder if we might go in a slightly different way: Instead of turning on and off the warnings in the tests multiple times with `rlang::local_options(rlib_warning_verbosity = "verbose|quiet")`, what if we set `rlang::local_options(rlib_warning_verbosity = "verbose")` at the top of the file (or we might need to do it in each `test_that` block, or we could use `old_options <- options(rlib_warning_verbosity = "verbose")` at the top and then `options(rlib_warning_verbosity = old_options)` at the end). This way we always get the warning in the test[1].

Then when we remove our `local_edition(2)` and `local_edition(3)` lines, we'll get a bunch of warnings in the output, but we can wrap the `expect_warning()` or `compare_dplyr_bdingin()` with `suppressWarnings(...)` which is the approved way to do this, it appears: https://github.com/r-lib/testthat/issues/1083

That way we don't have to worry about the order of the tests, or that we're not catching the warnings, etc. How does that sound?

[1] — it turns out because of the way that our `compare_dplyr_binding()` works, we can't just test the warnings once and then never again unless we cause the warning first outside of that.

File: r/tests/testthat/test-dplyr-summarize.R
Comment (alistaire47) [reply r/tests/testthat/test-dplyr-summarize.R:324]: I tried `default`, but in some cases (maybe consistent? but I don't want to promise that) the `.frequency = "once"` ones still threw a warning afterwards. Easier/safer to just make it verbose at the start of the file and back to default at the end; let me try it out

Comment (jonkeane) [reply r/tests/testthat/test-dplyr-summarize.R]: 🤔 I wonder if changing that setting actually resets the counter? 

Anyway, for our purposes, during tests, we probably do want it to always end up firing cause that's easier to catch + test or ignore.

Comment (alistaire47) [reply r/tests/testthat/test-dplyr-summarize.R]: Changed and pushed

**Comment by jonkeane** [comment]
Awesome, I've kicked off the CI again (sorry about the delay on that! Feel free to @ me when you need that approved again — it'll only be needed for this first PR!)

I know this wasn't in the original ticket, but would you mind also removing the lines that set the testthat edition: `local_edition(2)` and `local_edition(3)` https://github.com/apache/arrow/blob/18803eaafdea1994da42c5ea4bd8cf046057c357/r/tests/testthat/test-dplyr-summarize.R#L300-L302 https://github.com/apache/arrow/blob/18803eaafdea1994da42c5ea4bd8cf046057c357/r/tests/testthat/test-dplyr-summarize.R#L345 (and then there are at least two more later on...) and then add `SuppressWarnings()` around the tests that have multiple warnings?

File: r/tests/testthat/test-dplyr-summarize.R
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-summarize.R:21]: using `options` directly totally works, but since we have a call to `withr::local_options()` here already, does it work if we put `rlib_warning_verbosity = "verbose"` inside that list?

Comment (alistaire47) [reply r/tests/testthat/test-dplyr-summarize.R:21]: We're also raising 10 or so warnings here that 

```r
`summarise()` has grouped output by 'some_grouping'. You can override using the `.groups` argument.
```

Do we want to suppress those as well while we're at it?

Comment (jonkeane) [reply r/tests/testthat/test-dplyr-summarize.R:21]: Could we assert the message once and then wrap in `suppressMessages()` like we do for the warnings?

File: r/tests/testthat/test-dplyr-summarize.R
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-summarize.R:340]: OOOH, that's fantastic. I don't think I knew that `suppressWarnings()` would let you specify classes, that's really nice!

File: r/tests/testthat/test-dplyr-summarize.R
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-summarize.R:293]: I like that we're not repeating `suppressWarnings()` a bunch, but I wonder if it might be better in this case to wrap each `compare_dplyr_binding()` separately so it's a bit close to the code we know will be generating too many warnings for us. I'm on the fence and fine to do either way — totally up to you!

File: r/tests/testthat/test-dplyr-summarize.R
Comment (alistaire47) [reply r/tests/testthat/test-dplyr-summarize.R]: Wait we could set an option for this too instead. From `?dplyr::summarize`:

```
In addition, a message informs you of that choice, unless the result is ungrouped, the option "dplyr.summarise.inform" is set to FALSE, or when summarise() is called from a function in a package.
```

**Review by jonkeane — APPROVED** [review]
Looks good, one small comment that you can take or leave. I'll merge when the CI is green. 

Thanks for all this work!

File: r/tests/testthat/test-dplyr-summarize.R
Comment (alistaire47) [reply r/tests/testthat/test-dplyr-summarize.R:340]: Yeah, I think this was an R 4.0 thing. Ironically I was playing around with similar ideas at the same time in [rror](https://github.com/alistaire47/rror). Maybe there's still some value to that package, but I suspect it's mostly obsolete now. I should really write a blog post about the concepts, though; nobody classes conditions enough

File: r/tests/testthat/test-dplyr-summarize.R
Comment (jonkeane) [reply r/tests/testthat/test-dplyr-summarize.R]: Oh nice, yeah maybe do a local option there setting it to false with a comment as to why | what the message is that we are suppressing

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-versions

Running our version check nightly job based on https://github.com/apache/arrow/pull/12867#discussion_r849645179 to see if we might need to gate the class specification for a little while.

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-versions

File: r/tests/testthat/test-dplyr-summarize.R
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-summarize.R:23]:
```suggestion
  # This prevents the warning in `summarize()` about having grouped output without 
  # also specifying what to do with `.groups`
  dplyr.summarise.inform = FALSE
```

It might be nice to have a comment about why we need the `dplyr.summarise.inform=FALSE` here so we remember. The others are more-or-less straightforward, but that one I'm not sure I would remember what causes it + where.

**Review by jonkeane — APPROVED** [review]
Looks good, one small suggestion, and we can wait for the CI to go green

**Comment by jonkeane** [comment]
Hmm, it looks like `tryInvokeRestart` isn't available prior to 3.6: https://github.com/ursacomputing/crossbow/runs/6013703013?check_suite_focus=true#step:5:21063

https://github.com/wch/r-source/blob/7d25524f8b83d63979cc18f0914a321bfbe9a819/src/library/base/R/conditions.R#L86-L92 might be helpful if we wanted to approximate the old behavior. Or maybe we should just do something like this:

```
.suppressWarnings <- suppressWarnings
suppressWarnings <- function(exprs, ...) {
  if ({r version > 4.0.0}) {
    .suppressWarnings(exprs, ...)
  } else {
    .suppressWarnings(exprs)
  }
}
```

And then we don't have to worry about the internals too much?

**Comment by jonkeane** [comment]
We could also skip these tests on older versions if getting the suppress behavior is too onerous. We don't do it anymore, but when we had 3.3 support, we did it with median:

https://github.com/apache/arrow/pull/12072/files#diff-5315fa7158cd65ca9405299bf0b4551952e2555b82ea403204df2da1dc7b0013R283-L287

**Comment by alistaire47** [comment]
> Hmm, it looks like `tryInvokeRestart` isn't available prior to 3.6

Oops, we can just use `invokeRestart()`, which did exist, or backport `tryInvokeRestart()` too, or add [the backports package](https://github.com/r-lib/backports) to `Suggests` and use that (it's probably a recursive dependency already? but not sure). For now I went with the first option since it's the simplest.

File: r/tests/testthat/helper-arrow.R
Comment (jonkeane) [inline r/tests/testthat/helper-arrow.R:69]: Could we wrap this in an `if(getRversion() < "4.0.0") {...}` So that we can continue using the official `suppressWarnings()` (with the additions of tryInvokeRestart, etc.) for versions that support it?

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-versions

**Review by jonkeane — APPROVED** [review]
Almost getting there! I made one more minor suggestion for the the backport gating to make it slightly more robust.

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-versions

**Comment by alistaire47** [comment]
wait crap that has an extra `}`; fix incoming

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-versions

**Comment by jonkeane** [comment]
Ok, I'm just waiting for the rest of the CI to be green. The failure in the versions https://github.com/ursacomputing/crossbow/actions/runs/2164085819 is not due to this PR, that's there in the nightlies, but we _do_ see 3.6 and 3.5 both passing, which is fantastic!

**Pull request closed without merging**
